src/sinks/mod.rs

```diff
@@ -96,26 +96,27 @@ pub enum HealthcheckError {
 impl VectorSink {
     pub async fn run01<S>(self, input: S) -> Result<(), ()>
     where
-        S: futures01::Stream<Item = Event, Error = ()>,
+        S: futures01::Stream<Item = Event, Error = ()> + Send + 'static,
     {
-        match self {
-            Self::Futures01Sink(sink) => input.forward(sink).compat().map_ok(|_| ()).await,
-        }
+        self.run(input.compat()).await
     }
 
-    pub async fn run<S>(self, input: S) -> Result<(), ()>
+    pub async fn run<S>(mut self, input: S) -> Result<(), ()>
     where
-        S: futures::Stream<Item = Result<Event, ()>> + Send,
+        S: futures::Stream<Item = Result<Event, ()>> + Send + 'static,
     {
-        self.run01(Compat::new(input.boxed())).await
+        match self {
+            Self::Futures01Sink(sink) => input.forward(sink.sink_compat()).await,
+            Self::Stream(ref mut s) => s.run(Box::pin(input)).await,
```

Comment:
Can't this be just `pin_mut!` instead of `Box::pin`?

Reply:
I'm not sure that this is possible?
`run` signature:
```rust
async fn run(
        &mut self,
        input: Pin<Box<dyn futures::Stream<Item = Result<Event, ()>> + Send>>,
    ) -> Result<(), ()>;
```

I need `Box<dyn Trait>` because generic can not be used...
```bash
error[E0038]: the trait `sinks::util::sink::StreamSink2` cannot be made into an object
   --> src/sinks/mod.rs:71:12
    |
71  |     Stream(Box<dyn util::StreamSink2 + Send>),
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `sinks::util::sink::StreamSink2` cannot be made into an object
    | 
   ::: src/sinks/util/sink.rs:139:11
    |
139 | pub trait StreamSink2 {
    |           ----------- this trait cannot be made into an object...
140 |     async fn run<S: futures::Stream<Item = Result<Event, ()>> + Send>(
    |              --- ...because method `run` has generic type parameters
    |
    = help: consider moving `run` to another trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0038`.
error: could not compile `vector`.
```

We can use generic, but only if add `Self: Sized`, but then we it will not be possible call `run` method from `VectorSink` (because we have trait, not object) :disappointed:

Reply:
Ah, right. Well, this isn't temporary though, it is the aspect of the new design, right? I don't think it's problematic, but it would be interesting to see if there are workarounds for this...